### PR TITLE
#658 Add a profile picture editor

### LINKED ISSUE #658: Add a profile picture editor and use it upon profile picture upload
### Problem

For uploading both the group profile image and the user profile image, there are multiple problems:

- Pictures that are too large, wasting bandwidth (e.g. a 2MB picture can be uploaded, when at most the user will only ever see a tiny version of it)
- Pictures can be oblong in shape, resulting in oval user pictures

### Solution

- Find existing open source solutions to this problem (e.g. [here's an example](https://github.com/beakerbrowser/fritter/blob/master/com/crop-popup.js)) that use Canvas to edit the image
- Limit all profile pictures to some maximum Height x Width (e.g. 512x512)
- Fill in any gaps created by the scaling/resizing of the image
- Make sure the profile picture editor allows the user to shrink/enlarge and then crop the image to the section of the picture they uploaded, and then generate a new picture that will then be uploaded from that. This process is described in more detail below:

The end result is that all profile pictures (both group and user pictures) should be set to some constant values predefined in the app (e.g. let's say 512x512).

The point of the profile picture editor is to take as input [USER PICTURE] and produce as output => [CROPPED AND SCALED USER PICTURE OF SIZE 512x512], that is then the final picture that is actually uploaded to the server.

The profile picture editor needs to allow the user to both scale and move the picture, so that they can pick the section of the photo that they want to be their profile picture. It should also allow them to maybe select a background color to fill in any gaps (if there are any) that are created as a result of moving+scaling the photo in such a way that its edges are visible in the final cropped section.

EDIT: there is a related issue #1132 - which is necessary to close before implementing this issue. It's just about scaling the images - not about the UI for a profile picture editor.

### PR DESCRIPTION
**Pull request by SebinSong**
closes #658

### CODE REVIEW

**Review by taoeffect — COMMENTED** [review]
Excellent work @SebinSong! This is looking really great! 😄 

Some quick feedback on things I noticed:

- User can drag the image anywhere - it should be bounded by the image itself, right?
- When testing mobile in FFDevEdition and updating my profile picture (in responsive mobile mode), I wasn't able to drag the zoom sliders for some reason.

**Comment by SebinSong** [comment]
@taoeffect 
Updated the PR with the fix for those issues.

> When testing mobile in FFDevEdition and updating my profile picture (in responsive mobile mode), I wasn't able to drag the zoom sliders for some reason.

I did see this happening in Firefox devtool too but when I tried the same in chrome devtool everything works fine as you can see from below video.

https://user-images.githubusercontent.com/17641213/197938691-0b2d186e-757f-47a0-bd42-b02e5ec2eec0.mp4

More importantly, I tried it on my android phone (on chrome browser via the browsersync external link- http://172.30.1.56:3000), It did work fine there too. It seems like a bug on firefox dev edition itself. :)

Thanks,

File: frontend/views/components/avatar-editor/AvatarEditorModal.vue [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+      button.is-outlined(
+        type='button'
+        @click='close'
+      ) Cancel
+
+      button-submit(
+        key='save'
+        data-test='saveBtn'
+        @click='submit'
+      ) Save
```
Comment (taoeffect) [inline frontend/views/components/avatar-editor/AvatarEditorModal.vue]: Missing i18n, try with `tag='button'` and `tag='button-submit'`, or include a child `i18n` element

**Review by taoeffect — CHANGES_REQUESTED** [review]
Very clean and nicely done @SebinSong!

Minor comments-  also, it still allows me to drag the images beyond their borders / outside of their own bounding box.

Screen Shot 2022-10-26 at 2 13 24 PM

**Comment by SebinSong** [comment]
@taoeffect  thanks for review. 

Thought I would need a help with clarifying what the below requirement means.
> it still allows me to drag the images beyond their borders / outside of their own bounding box.

Does this mean that the the clipping circle should never leave the boundary of the image? For example,

 

The red boundary is the bounding-box of the image on the canvas, whose size can change as the user zooms in & out of the image. and when the user moves the image around, Does the movement(or translation) of the image have to be limited to the points where the clipping circle never goes beyond the four edges of the red-boundary here?
To be detailed,


 

This is not ok because the image has been moved out of the clipping circle. and **the correct behaviour** would be,

 

The moving of image stops at this point, where the edges of the images are flush with the edges of the clipping circle.

Is this correct?
Hope my drawings here are being helpful here. :)

**Comment by SebinSong** [comment]
@taoeffect 
here is the video demonstration of what I commented above.


https://user-images.githubusercontent.com/17641213/198231509-85525a99-cf6a-4e29-b92b-4ae0b43d2b1f.mp4


Is this the behaviour we want? cheers,

**Comment by taoeffect** [comment]
@SebinSong Exactly right! 👍

**Comment by SebinSong** [comment]
@taoeffect updated the PR with the work for the change request.

**Review by taoeffect — APPROVED** [review]
Brilliant! 🥇 😄 🎉

**Pull request merged**